### PR TITLE
Scheme/Code bug fixes

### PR DESCRIPTION
**Pull request by as2388**

### CODE REVIEW

File: core_data_modules/data_models/scheme.py [hunk 1]
```diff
@@ -34,40 +38,43 @@ def get_code_with_control_code(self, control_code):
...
+            documentation = {
+                "URI": validators.validate_string(doc_map["URI"])
+            }
         
-        return scheme
+        return cls(scheme_id, name, version, codes, documentation)
     
     def to_firebase_map(self):
-        ret = {}
+        ret = dict()
```
Comment (lukechurch) [inline core_data_modules/data_models/scheme.py]: After doing this for a lot of time, I don't think this really makes sense, in this case it's very obvious from context that this is a dictionary not a set.

I suggest that we standardise on using x = {} only for dictionaries, and use the explicit form for handling sets.

File: core_data_modules/data_models/scheme.py [hunk 2]
```diff
@@ -34,40 +38,43 @@ def get_code_with_control_code(self, control_code):
 
     def get_code_with_match_value(self, match_value):
         for code in self.codes:
-            if match_value in code.match_values:
+            if code.match_values is not None and match_value in code.match_values:
```
Comment (lukechurch) [inline core_data_modules/data_models/scheme.py]: I would prefer to move an assert that match_value was not none to a precondition after the first line - it doesn't seem to make sense to me to call this method with match_value == None

File: core_data_modules/data_models/scheme.py [hunk 3]
```diff
@@ -2,19 +2,23 @@
...
+        self.documentation = documentation
 
     def __eq__(self, other):
         if not isinstance(other, self.__class__):
             return False
-        return other.scheme_id == self.scheme_id & \
-            other.name == self.name & \
-            other.version == self.version & \
-            other.documentation == self.documentation & \
+        return other.scheme_id == self.scheme_id and \
```
Comment (lukechurch) [inline core_data_modules/data_models/scheme.py]: Is this a bug fix, or a style change?

File: core_data_modules/data_models/scheme.py [hunk 4]
```diff
@@ -2,19 +2,23 @@
 
 
 class Scheme(object):
-    scheme_id = None
-    name = None
-    version = None
-    codes = []
-    documentation = dict()
+    def __init__(self, scheme_id, name, version, codes, documentation=None):
+        if documentation is None:
```
Comment (lukechurch) [inline core_data_modules/data_models/scheme.py]: Why not default it to a dictionary?

File: core_data_modules/data_models/scheme.py (hunk 3 above)
Comment (as2388) [reply core_data_modules/data_models/scheme.py]: Bug fix. As explained yesterday, & takes precedence over equals. If you wanted to keep the (in my opinion not Pythonic) &s you'd need to add brackets

File: core_data_modules/data_models/scheme.py (hunk 4 above)
Comment (as2388) [reply core_data_modules/data_models/scheme.py]: As discussed yesterday, because then the same dictionary would be shared across all scheme instances.

Comment (lukechurch) [reply core_data_modules/data_models/scheme.py]: So to be clear  
```
def __init__(self, scheme_id, name, version, codes, documentation={}):
```

Shares the generated dictionary across all instances?

File: core_data_modules/data_models/scheme.py (hunk 2 above)
Comment (lukechurch) [reply core_data_modules/data_models/scheme.py]: Oh, sorry, yeah, it's inside the for loop, ignore.

File: core_data_modules/data_models/scheme.py (hunk 4 above)
Comment (as2388) [reply core_data_modules/data_models/scheme.py]: Yes (eww)

File: core_data_modules/data_models/scheme.py (hunk 1 above)
Comment (lukechurch) [reply core_data_modules/data_models/scheme.py]: I don't agree, I think this form is unpythonic. I don't see any risk -

```

File: core_data_modules/data_models/scheme.py (hunk 3 above)
Comment (lukechurch) [reply core_data_modules/data_models/scheme.py]: ack

File: core_data_modules/data_models/scheme.py (hunk 1 above)
Comment (as2388) [reply core_data_modules/data_models/scheme.py]: I'm fine if it's tested. But the lack of tests for this file have caused an afternoon of pain already; this felt like an extra safety net. 

As a (possibly) side note: this is the first piece of untested code we have allowed in Core, and I'm reluctant to ever let that happen again, even if there were claims of informal testing having worked.

Comment (lukechurch) [reply core_data_modules/data_models/scheme.py]: >I'm fine if it's tested. But the lack of tests for this file have caused an afternoon of pain already; this felt like an extra safety net.

I need to be persuaded that it's ok for code to be more reasonable if it's been tested.

> As a (possibly) side note: this is the first piece of untested code we have allowed in Core, and I'm reluctant to ever let that happen again, even if there were claims of informal testing having worked.

IMHO this is not an appropriate venue for this discussion

**Comment by as2388** [comment]
Merging. File an issue if you want to correct dict -> {} (it would need to be done across the repo anyway, not just in this PR)

**Pull request merged**